File: Solutions/Task2.cs
namespace Task2;

public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution
{
    /*
    Сложность.
    Пусть N будет количеством узлов в дереве. 
    
    Временная сложность: O(N).
    Это связано с тем, что в рекурсивной функции longestPath мы входим и выходим из каждого узла только один раз. 
    Мы знаем это, потому что каждый узел идет от своего родителя, 
    а в дереве у узлов есть только один родитель. 
    
    Пространственная сложность: O(N). 
    Пространственная сложность зависит от размера нашего неявного стека вызовов во время обхода DFS, 
    который связан с высотой дерева. 
    В худшем случае дерево скошено, поэтому высота дерева равна O(N). 
    Если дерево сбалансировано, сложность будет O(logN).

    */


    private int diameter;
    public int DiameterOfBinaryTree(TreeNode root)
    {
        diameter = 0;
        longestPath(root);
        return diameter;
    }
    private int longestPath(TreeNode node)
    {
        if (node == null) return 0;
        // рекурсивно находим самый длинный путь в
        // и левый дочерний элемент, и правый дочерний элемент
        int leftPath = longestPath(node.left);
        int rightPath = longestPath(node.right);

        // обновляем диаметр, если leftPath плюс rightPath больше
        diameter = Math.Max(diameter, leftPath + rightPath);

        // возвращаем самый длинный между leftPath и rightPath;
        // добавляем 1 для пути, соединяющего узел и его родителя
        return Math.Max(leftPath, rightPath) + 1;
    }
}

File: Solutions/Task2.py
 class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

    # Сложность.
    # Пусть N будет количеством узлов в дереве. 
      
    # Временная сложность: O(N).
    # Это связано с тем, что в рекурсивной функции longest_path мы входим и выходим из каждого узла только один раз. 
    # Мы знаем это, потому что каждый узел идет от своего родителя, 
    # а в дереве у узлов есть только один родитель. 
      
    # Пространственная сложность: O(N). 
    # Пространственная сложность зависит от размера неявного стека вызовов во время обхода DFS, 
    # который связан с высотой дерева. 
    # В худшем случае дерево скошено, поэтому высота дерева равна O(N). 
    # Если дерево сбалансировано, сложность будет O(logN).


class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        diameter = 0

        def longest_path(node):
            if not node:
                return 0
            nonlocal diameter

            left_path = longest_path(node.left)
            right_path = longest_path(node.right)

            diameter = max(diameter, left_path + right_path)

            return max(left_path, right_path) + 1

        longest_path(root)
        return diameter